### ISSUE TITLE
Improve Test Runner error experience

### ISSUE THREAD

**Issue by jennifer-shehane** [cypress-io/cypress#3762]
# Goal

The goal is to make the process of debugging tests and code easier.

# Why
1. Spending less time debugging and deciphering errors means more time for users to do what they _want_ to be doing (e.g. writing more tests, writing more code, stopping writing code because they’ve achieved their goals).
2. When users can solve their own problems, this reduces the time our team spends on support.
3. Better errors makes software easier to use for everyone, especially beginners or non-traditional users who might not have context on how to debug issues outside of what error messages instruct them to do. 

# Implementation

## Types of errors

![](https://user-images.githubusercontent.com/7514489/53921702-ff785a80-4036-11e9-8140-53505c3773f9.png)

## Current design of error within Command Log

![error-design-7](https://user-images.githubusercontent.com/1271364/54336021-3a8f1500-4659-11e9-9ad2-eb3513494546.png)

## Current design applied to types of errors

<img width="1427" alt="error-designs-8" src="https://user-images.githubusercontent.com/1271364/54340712-5ea52300-4666-11e9-83ae-f972e488aa71.png">

### Cypress Error (one code frame)

```js
{
  name: "CypressError",
  message: "", // Error message (logged in console, used in run mode)
  stack: "",
  mdMessage: "", // Error message with escaped markdown (used in GUI)
  codeFrames: [
    {
      frame: "", // Code frame string
      filePath: "" // Path to open file
    }
  ],
  docs: "", // Either full URL or path after on.cypress.io
  template: "cypress-single-frame" // Template type, used to compose full error
}
```

### Cypress Error (two code frames)

```js
{
  name: "CypressError",
  message: "",
  stack: "",
  mdMessage: "",
  codeFrames: [
    { // Code frame to be displayed first, likely the spec code
      frame: "",
      filePath ""
    },
    { // Code frame to be displayed second, such as the plugins code
      frame: "",
      filePath ""
    }
  ],
  docs: "",
  template: "cypress-multiple-frames"
}
```

### Assertion Error

```js
{
  name: "AssertionError",
  message: "",
  stack: "",
  expected: "",
  actual: "",
  diff: "",
  codeFrames: [
    {
      frame: "",
      filePath: ""
    }
  ],
  docs: "",
  template: "assertion"
}
```

### Spec/Application Error

```js
{
  name: "",
  message: "",
  stack: "",
  codeFrames: [
    {
      frame: "",
      filePath: ""
    }
  ],
  docs: "",
  template: "other"
}
```

# Existing Issues

There are a number of existing issues that might make sense to address under the umbrella of “error improvement" (I’m sure there are more): 

* Add retry to `baseUrl` not present warning https://github.com/cypress-io/cypress/issues/3284
* Better messaging when failing by type https://github.com/cypress-io/cypress/issues/2050
* Failed assertion UI in open mode https://github.com/cypress-io/cypress/issues/3129
* Module API doesn’t fail the promise with Cypress error message https://github.com/cypress-io/cypress/issues/3216
* Root level `after` hooks don’t error https://github.com/cypress-io/cypress/issues/3348
* Improve printing error messages in Test Runner https://github.com/cypress-io/cypress/issues/1664
* better stack trace/ sourceMaps https://github.com/cypress-io/cypress/issues/881
* asterisks appearing in command log https://github.com/cypress-io/cypress/issues/1360
* Add support for backticks in command log https://github.com/cypress-io/cypress/issues/3608
* Exceptions in global 'after' are not shown correctly in UI https://github.com/cypress-io/cypress/issues/3904
* Exceptions in after() don't show up in UI https://github.com/cypress-io/cypress/issues/3348
* Exception in root level after() doesn't fail test run in interactive mode https://github.com/cypress-io/cypress/issues/2296
* Improve uncaught errors https://github.com/cypress-io/cypress/issues/1761
* TypeError: Cannot read property 'replace' of undefined" at Object.appendErrMsg https://github.com/cypress-io/cypress/issues/1669
* .should('deep.equal') does not show a full diff comparison on failures https://github.com/cypress-io/cypress/issues/4084
* err when cy.visit('file://..') https://github.com/cypress-io/cypress/issues/4452
* Show warning message in command log https://github.com/cypress-io/cypress/issues/4917
* Add warning to command log for 'uncaught exception' https://github.com/cypress-io/cypress/issues/4933
* Capture stack traces from spec frame https://github.com/cypress-io/cypress/issues/4932

**Comment by scharf** [comment]
see Stack overflow question [How to figure out in which file and code line a cypress test has failed?](https://stackoverflow.com/q/55493100/2297345)

**Comment by tedpennings** [comment]
I would love to see the current browser URL recorded in Cypress assertion errors and appear in the test error output to stdout/stderr. Some of our URLs are very long and with important IDs cutoff in screenshots/video. It would be easier to investigate errors from CI (non-interactive) with URLs logged. (Locally this isn't a problem.) Thank you!

**Comment by brian-mann** [comment]
Latest feedback from @chrisbreiding 	and me pairing...

* add finder icon to open in finder
* add editor icon to consistently indicate to user how to open in IDE
* add editor icon next to the test title on hover to open the test definition in the IDE
* add a horizontal bar at the top of the tests to indicate which spec file it is
* list of editors based on what was detected…
    * error handling
    * checkbox for “use system default”
    * on cypress open, check to see that editor is available else blow away the cached value
* View Stack Trace | Print | Copy
* when clicking view stack trace - rather than open in IDE, this should open in devtools
* use console.group() better to connect actions in the command log to the console.log
    * printing key/values or other messages
    * Test: “<testname>”
    * Error: ‘….’
* Style the “View stack trace” expansion
* Hover over lines in the codeframe to make them clickable

**Comment by chrisbreiding** [comment]
More ideas for improvements from discussion with Brian on 1/17:

- Add learn more link to docs in modal too (like in desktop-gui)
- Output code frame in run mode stdout
- Don't link cypress:// error files in stack trace
- Send more data to dashboard (server/lib/reporter.coffee)
    - full error data with code frame, source mapped stack, etc
    - source mapped body
    - ask Jeff

**Comment by jennifer-shehane** [comment]
This was partially released in 4.3.0. The full release with code frames is still in progress at https://github.com/cypress-io/cypress/pull/3930
